### ISSUE TITLE
Return the end of the number

### ISSUE THREAD

**Issue by lemire** [CarlVerret/csFastFloat#61]
The C++ fastfloat library returns the end of the number (e.g., given `132.32 ...`, it will return a pointer to the space).

The csFastFloat is a port of the C++ fastfloat library, and it should therefore be able to provide the same capability without much complexity.

**Comment by gfoidl** [comment]
In .NET APIs for this use a `out int consumed` argument. So the API would like

```diff
public static unsafe double ParseDouble(string s, chars_format expectedFormat = chars_format.is_general, char decimal_separator = '.');
public static unsafe double ParseDouble(ReadOnlySpan<char> s, chars_format expectedFormat = chars_format.is_general, char decimal_separator = '.');
+ public static unsafe double ParseDouble(ReadOnlySpan<char> s, out int consumed, chars_format expectedFormat = chars_format.is_general, char decimal_separator = '.');
public static unsafe double ParseDouble(ReadOnlySpan<byte> s, chars_format expectedFormat = chars_format.is_general, byte decimal_separator = (byte)'.');
+ public static unsafe double ParseDouble(ReadOnlySpan<byte> s, out int consumed, chars_format expectedFormat = chars_format.is_general, byte decimal_separator = (byte)'.');
```

That way a consumer can slice / advance it's input span as needed after a call to `ParseDouble`.

Note: adding the out-argument to the existing overload is technically a binary breaking change.
(I would not have a problem to take this break -- if documented in the releases -- as the existing overload will become less meaningful once the new ones are added).

For float analogous.

**Comment by lemire** [comment]
@gfoidl At a glance, it does seem like I implemented what you proposed in https://github.com/CarlVerret/csFastFloat/pull/64, did I not?

> adding the out-argument to the existing overload is technically a binary breaking change.

Can you elaborate ? Would it break existing code ? If so, how? (I am not arguing. I just want to understand.)

**Comment by gfoidl** [comment]
#64 is fine, and matches the proposal (except `int` <-> `long`).

> Would it break existing code ?

You added a new overload, so it's fine and nothing will break.

If the out-argument would have been augmented to the existing method, then it's a breaking change.
Customers would need to re-compile with the new library to avoid runtime-failures, that would occur if just the library is updated (say copy & pasted the DLL).
In the .NET-ecosystem versioning is done according semver.org, so a breaking change is allowed iif the major-version is updated.

**Comment by lemire** [comment]
We definitively should bump the version. In fact, I would go straight to 2.0. We are making a lot of changes.

cc @CarlVerret

**Comment by gfoidl** [comment]
👍 

Then 
https://github.com/CarlVerret/csFastFloat/blob/fd0c32ca3dcec0f5a2f155a3eb41652f85a78887/csFastFloat/FastDoubleParser.cs#L56 
https://github.com/CarlVerret/csFastFloat/blob/fd0c32ca3dcec0f5a2f155a3eb41652f85a78887/csFastFloat/FastDoubleParser.cs#L81
https://github.com/CarlVerret/csFastFloat/blob/fd0c32ca3dcec0f5a2f155a3eb41652f85a78887/csFastFloat/FastDoubleParser.cs#L175
(so the methods with out-argument for the consumded elements) could be dropped, as they don't any additional value and just "bloat" the API-surface. Any potential consumer can just discard `ParseDouble(..., out _, ...)` the argument, and the JIT will eliminate any usage.

**Comment by lemire** [comment]
@gfoidl 

Would `x = FastDoubleParser.ParseDouble(l);` still work given that `l` is a string? That's what we want.

**Comment by gfoidl** [comment]
Oh, no.

So please ignore my last comment, which would result in `x = FastDoubleParser.ParseDouble(l, out _);`.

**Comment by lemire** [comment]
@gfoidl Is there a way in C# to make `out` parameters optional?

**Comment by gfoidl** [comment]
No. The language requires that ref or out parameters don't have a default value, which IMHO makes sense, because what value would one return here? Any consumer of the method would be fooled about the actual value.

The overload without the `out` parameter could be a thin wrapper around the other one, like:
```c#
public static unsafe double ParseDouble(string s, chars_format expectedFormat = chars_format.is_general, char decimal_separator = '.')
  => ParseDouble(s, out _, expectedFormat, decimal_separator);
```
to avoid some repetetion of code.
